### PR TITLE
Nullability fixes in enum mapping

### PR DESCRIPTION
**Pull request by MaceWindu**

### CODE REVIEW

**Comment by jogibear9988** [comment]
what does not work?

**Comment by MaceWindu** [comment]
Check added tests (you can try run them without fixes and see results) - there were several issues with nullable enum fields and null in mappings:
1. duplicate mapping values exception when enum mapped to nullable value type (e.g. int? ) and have 0 and null in mapped values
2. default(type) instead of default(type?) used as default value for nullable enum mapped to nullable value type
3. in general if enum mapped to value type had mapping to null - it produced exception

With this change mapping to value type with null involved behaves same as mapping to reference type (e.g. string)

**Comment by ili** [comment]
why 1.9.0? you suppose it is not ready to merge?

File: Source/Common/ConvertBuilder.cs [hunk 1]
```diff
@@ -627,7 +638,7 @@ select a
 					?? mappingSchema.GetDefaultFromEnumType(typeof(Enum))
 					?? Enum.GetUnderlyingType(type);
 
-			if (enumType.IsNullable() && !defaultType.IsClassEx() && !defaultType.IsNullable())
+			if ((enumType.IsNullable() || fields.Any(attrs => attrs.Count != 0 && attrs[0].Value == null)) && !defaultType.IsClassEx() && !defaultType.IsNullable())
```
Comment (sdanyliv) [inline Source/Common/ConvertBuilder.cs]: Can you describe this change?

**Comment by MaceWindu** [comment]
I assigned only urgent defects and already approved fixes to 1.8.3. If we will have anything else ready to release with it - I have no objection too.

File: Source/Common/ConvertBuilder.cs (hunk 1 above)
Comment (MaceWindu) [inline Source/Common/ConvertBuilder.cs]: good question... I think I need to add more tests to see if it introduce change of behavior for some mapping cases.

As for this change it fixes following cases:

enum->db mapping for following case:
```cs
enum Enum
{
// explicit mapping to null
    [MapValue(null)] Value
}

// non-nullable field
Enum field;

// old behavior - mapped to default(int) = 0
// new behavior - mapped to default(int?) = null
```

and for db-> enum mapping it fixes exception for this case:

```cs
enum Enum
{
    [MapValue(0)] Value1,
    [MapValue(null)] Value2
}
```
```
Mapping ambiguity. MapValue(0) attribute is defined for both 'Enum.Value1' and 'Enum.Value2'.
```

**Comment by MaceWindu** [comment]
Ok, I've reviewed enum mapping functionality once more and don't see issues with this PR.
As a bonus - added xml comments to MapValueAttribute as a part of our documentation efforts

**Pull request merged**